feat(function): Implement map_delete function

shamb0 opened the pull request:
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

Implement `map_delete` function

- Addresses part of [15295](https://github.com/datafuselabs/databend/issues/15295)

## Tests

- [x] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/datafuselabs/databend/15480)
<!-- Reviewable:end -->

shamb0 commented:
Hi @b41sh,

- I've completed the first draft of map_delete. Could you please review the code and provide feedback when you get a chance? 

- Let me know if there are any areas that need improvement or alternative approaches to consider.

Thanks!

b41sh commented:
> Hi @b41sh,
> 
> * I've completed the first draft of `map_delete`. Could you please review the code and provide feedback when you get a chance?
> * Let me know if there are any areas that need improvement or alternative approaches to consider.
> 
> Thanks!

Hi @shamb0 , thanks for continuing to contribute code.
Since the `map_delete` function has a variable number of arguments and may have more than one key argument, it is better to use Factory function instead of Builtin function. You can refer to these Factory functions as examples.

https://github.com/datafuselabs/databend/blob/main/src/query/functions/src/scalars/string_multi_args.rs#L88
https://github.com/datafuselabs/databend/blob/main/src/query/functions/src/scalars/string_multi_args.rs#L247
https://github.com/datafuselabs/databend/blob/main/src/query/functions/src/scalars/geo.rs#L311

There are many more implementations of the factory function, and you can find more examples by searching the code base for `registry.register_function_factory`

shamb0 commented:
> > Hi @b41sh,
> > 
> > * I've completed the first draft of `map_delete`. Could you please review the code and provide feedback when you get a chance?
> > * Let me know if there are any areas that need improvement or alternative approaches to consider.
> > 
> > Thanks!
> 
> Hi @shamb0 , thanks for continuing to contribute code. Since the `map_delete` function has a variable number of arguments and may have more than one key argument, it is better to use Factory function instead of Builtin function. You can refer to these Factory functions as examples.
> 
> https://github.com/datafuselabs/databend/blob/main/src/query/functions/src/scalars/string_multi_args.rs#L88 https://github.com/datafuselabs/databend/blob/main/src/query/functions/src/scalars/string_multi_args.rs#L247 https://github.com/datafuselabs/databend/blob/main/src/query/functions/src/scalars/geo.rs#L311
> 
> There are many more implementations of the factory function, and you can find more examples by searching the code base for `registry.register_function_factory`
> 
> The factory function is not written in the same way as the builtin function, it can support more flexible argument types, but it needs to deal with more problems, there are mainly the following steps:
> 
> First of all, you need to check whether the input arguments are supported types, for the `map_delete` function, the number of arguments is at least two, the type of the first argument must be map, the type of the second and later arguments must be the same as the type of the map key, and can not be nullable. If the arguments do not meet these conditions, return `NULL`, otherwise, we can get `args_type` and `return_type`, used to generate `FunctionSignature`.
> 
> The second part of writing the logic for executing the function is that the input args are of type `ValueRef`, and each argument could be a `Scalar` or a `Column`, which needs to be handled in a different way. You also need a builder to store the result data, there are many examples in this part, you can refer to the implementation of other functions.

Thanks @b41sh, for sharing those details! I'll take a good look at the requirements and circle back with any questions or updates.

b41sh commented:
Hi @shamb0 It's been a long time since last update, have you made any progress recently?

shamb0 commented:
> Hi @shamb0 It's been a long time since last update, have you made any progress recently?

Hi @b41sh,

Sorry for the delay. I've been tied up with some personal matters. I have some updates that are partially complete and will aim to make a full commit in the next few days.

shamb0 commented:
Hi @b41sh,

I've refactored the implementation to use a factory function instead of a built-in function. The proto implementation is ready for your review. Currently, it only supports the key variant of type `DataType::String`. Since the `DataType` collection is extensive, could you please suggest which other variants we should support next?

b41sh commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -245,3 +310,86 @@ pub fn register(registry: &mut FunctionRegistry) {
+        _ => None,
+    });
+
+    let mut output_map_builder = ColumnBuilder::EmptyMap { len: 0 };
```
we can use the `return_type` to create the builder.
```rust
let mut builder = ColumnBuilder::with_capacity(&return_type, len);
```
`return_type` is same as `args_type[0]`, we can use `args_type` here if modify the `eval` as following:
```rust
eval: Box::new(move |args, ctx| {
```

shamb0 replied:
Changes implemented. Ready for your review.

b41sh commented:
> Hi @b41sh,
> 
> I've refactored the implementation to use a factory function instead of a built-in function. The proto implementation is ready for your review. Currently, it only supports the key variant of type `DataType::String`. Since the `DataType` collection is extensive, could you please suggest which other variants we should support next?

Hi @shamb0 Thank you for your modification. There are still some minor problems that need to be modified. You can have a look at them when you are free.

shamb0 commented:
> > Hi @b41sh,
> > I've refactored the implementation to use a factory function instead of a built-in function. The proto implementation is ready for your review. Currently, it only supports the key variant of type `DataType::String`. Since the `DataType` collection is extensive, could you please suggest which other variants we should support next?
> 
> Hi @shamb0 Thank you for your modification. There are still some minor problems that need to be modified. You can have a look at them when you are free.

Hi @b41sh,

Thank you for the review comments. I'll incorporate the suggested changes and have an updated version ready for your review soon.

shamb0 commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +244,56 @@ pub fn register(registry: &mut FunctionRegistry) {
+            return None;
+        }
+
+        let (key_type, value_type) = match arg_types.first()? {
```
@b41sh Could you clarify how we should handle `DataType::EmptyMap`? Since there’s nothing to process, can we simply `return None`?

shamb0 replied:
Changes implemented. Ready for your review.

b41sh replied:
we need to support case the first argument is `DataType::EmptyMap`, and the result is also an `EmptyMap`, for example
```sql
select map_delete({}, 'a', 'b');
```

shamb0 commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -245,3 +310,86 @@ pub fn register(registry: &mut FunctionRegistry) {
+    };
+
+    for idx in 0..(input_length.unwrap_or(1)) {
+        let (input_map_type, input_map) = match &args[0] {
```
Changes implemented. Ready for your review.

shamb0 commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -245,3 +310,86 @@ pub fn register(registry: &mut FunctionRegistry) {
+
+    let delete_key_list = if args.len() == 2 {
+        match &args[1] {
+            ValueRef::Scalar(ScalarRef::Array(keys_to_delete)) => {
```
Changes implemented. Ready for your review.

shamb0 commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +244,56 @@ pub fn register(registry: &mut FunctionRegistry) {
+                return_type: DataType::Map(Box::new(DataType::Tuple(vec![key_type, value_type]))),
+            },
+            eval: FunctionEval::Scalar {
+                calc_domain: Box::new(|_, _| FunctionDomain::Full),
```
Changes implemented. Ready for your review.

b41sh replied:
The domain can write like this, because the result have the same domain as the first argument.
```rust
calc_domain: Box::new(|_, args_domain| FunctionDomain::Domain(args_domain[0].clone()))
```

shamb0 commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +244,56 @@ pub fn register(registry: &mut FunctionRegistry) {
+                    ]))),
+                    DataType::Array(Box::new(key_array_type)),
+                ],
+                return_type: DataType::Map(Box::new(DataType::Tuple(vec![key_type, value_type]))),
```
Changes implemented. Ready for your review.

shamb0 commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +244,56 @@ pub fn register(registry: &mut FunctionRegistry) {
+            }
+        };
+
+        let key_array_type = match arg_types.get(1)? {
```
Changes implemented. Ready for your review.

shamb0 commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +244,56 @@ pub fn register(registry: &mut FunctionRegistry) {
         |map, _| map.len() as u64,
     );
 
+    registry.register_2_arg::<EmptyMapType, EmptyArrayType, EmptyMapType, _, _>(
```
Changes implemented. Ready for your review.

shamb0 commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +244,56 @@ pub fn register(registry: &mut FunctionRegistry) {
+    );
+
+    registry.register_function_factory("map_delete", |_, arg_types| {
+        if arg_types.len() != 2 {
```
Changes implemented. Ready for your review.

b41sh commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +241,125 @@ pub fn register(registry: &mut FunctionRegistry) {
+
+        for arg_type in args_type.iter().skip(1) {
+            match arg_type {
+                DataType::Array(box key_type) => {
```
We follow the snowflake syntax, the argument types should be same as the key type, not an array type.
Please refer to this document:
https://docs.snowflake.com/en/sql-reference/functions/map_delete#syntax

shamb0 replied:
Thankyou. Yes, I realised it latter.

b41sh replied:
you can write like this, `map_key_type` is `None` if the first argument is `EmptyMapType`.
```rust
if let Some(key_type) = map_key_type {
    for arg_type in args_type.iter().skip(1) {
        if arg_type != &key_type {
            return None;
        }
    }
} else {
    let key_type = args_type[1];
    if !key_type.is_boolean()
        && !key_type.is_string()
        && !key_type.is_numeric()
        && !key_type.is_decimal()
        && !key_type.is_date_or_date_time() {
    }
    for arg_type in args_type.iter().skip(2) {
        if arg_type != &key_type {
            return None;
        }
    }
}
```

b41sh commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +241,125 @@ pub fn register(registry: &mut FunctionRegistry) {
+                            _ => unreachable!(),
+                        };
+
+                        let mut filtered_kv_builder = ColumnBuilder::with_capacity(
```
This builder can be removed, only if the argument is a map need a builder, if the value is `Null` or `EmptyMap`, we can use 
```rust
output_map_builder.push_default();
```

shamb0 replied:
Good One, Thankyou

b41sh commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +241,125 @@ pub fn register(registry: &mut FunctionRegistry) {
+                            ScalarRef::Null | ScalarRef::EmptyMap => {
+                                filtered_kv_builder.push_default();
+                            }
+                            ScalarRef::Map(_) => {
```
We can create builder here

```rust
ScalarRef::Map(col) => {
    let mut filtered_kv_builder = ColumnBuilder::with_capacity(&col.data_type(), col.len());
    ..
```

shamb0 replied:
Good One, Thankyou

shamb0 replied:
Hi @b41sh,

I've addressed all the comments and made the necessary changes. 
The PR is now ready for In-take review.

Please let me know if there's anything else that needs attention.

Thanks,

b41sh replied:
create `delete_key_list` here like this
```rust
let mut delete_key_list = HashSet::new();
for j in 1..args.len() {
    let key = match &args[j] {
        ValueRef::Scalar(scalar) => scalar.clone(),
        ValueRef::Column(col) => unsafe { col.index_unchecked(idx) },
    };
    delete_key_list.insert(key);
}
```

shamb0 commented:
Hi @b41sh,

I've addressed all the comments and made the necessary changes.
The PR is now ready for In-take review.

Please let me know if there's anything else that needs attention.

Thanks,

b41sh commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +241,123 @@ pub fn register(registry: &mut FunctionRegistry) {
+                    let mut output_map_builder =
+                        ColumnBuilder::with_capacity(&return_type, input_length.unwrap_or(1));
+
+                    let delete_key_list = prepare_delete_key_list(args, input_length);
```
this should be removed, we need to get `delete_key_list` for each row by the index.

shamb0 replied:
I wanted to clarify one point regarding the suggested solution. The `key` is currently considered as `HashSet<ScalarRef::Array(ArrayColumn<T>)>`, but for processing the map_delete operation, we need `HashSet<T>`. Would it be acceptable to keep it as `HashSet<T>`, or should I modify the logic to use `HashSet<ScalarRef::Array(ArrayColumn<T>)>` instead?

b41sh commented on tests/sqllogictests/suites/query/functions/02_0074_function_map.test:
```diff
@@ -135,6 +135,63 @@ SELECT MAP_CAT(
+
+query
+SELECT 
+  MAP_DELETE(src_map, ['a_k2', 'b_k3'])
```
The syntax is not correct, we should support `MAP_DELETE(src_map, 'a_k2', 'b_k3')`, not an array.

shamb0 replied:
Ah, Thank you. I understand the reasoning behind your proposed code pattern for filtering the keys now. The necessary changes have been made. Please let me know if any additional changes are needed.

b41sh commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +241,123 @@ pub fn register(registry: &mut FunctionRegistry) {
+            return None;
+        }
+
+        let (map_key_type, _map_value_type) = match args_type[0].remove_nullable() {
```
`_map_value_type` can be removed, we can write like this:
```rust
let map_key_type = match args_type[0].remove_nullable() {
    DataType::Map(box DataType::Tuple(type_tuple)) => {
        if type_tuple.len() == 2 {
            Some(type_tuple[0].clone())
        } else {
            return None;
        }
    }
    DataType::EmptyMap => None,
    _ => {
        return None;
    }
};
```

b41sh commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +240,129 @@ pub fn register(registry: &mut FunctionRegistry) {
+                        ColumnBuilder::with_capacity(&return_type, input_length.unwrap_or(1));
+
+                    for idx in 0..(input_length.unwrap_or(1)) {
+                        let mut delete_key_list = HashSet::new();
+
+                        for input_key_item in args.iter().skip(1) {
+                            let input_key = match &input_key_item {
+                                ValueRef::Scalar(scalar) => scalar.clone(),
+                                ValueRef::Column(col) => unsafe { col.index_unchecked(idx) },
+                            };
+
+                            delete_key_list.insert(input_key.to_owned());
+                        }
+
+                        let input_map_sref = match &args[0] {
+                            ValueRef::Scalar(map) => map.clone(),
+                            ValueRef::Column(map) => unsafe { map.index_unchecked(idx) },
+                        };
+
+                        let inner_builder_type = match input_map_sref.infer_data_type() {
+                            DataType::Map(box typ) => typ,
+                            DataType::EmptyMap => DataType::EmptyMap,
+                            _ => unreachable!(),
+                        };
```
those codes can be moved to line 334, as if the value is `Null` or `EmptyMap`, we don't need build `delete_key_list`.

b41sh commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +240,129 @@ pub fn register(registry: &mut FunctionRegistry) {
+                            ScalarRef::Map(_) => {
+                                let mut filtered_kv_builder = ColumnBuilder::with_capacity(
+                                    &inner_builder_type,
+                                    input_length.unwrap_or(1),
```
We can use `col.len()` in `ScalarRef::Map(col)` as the capacity, not the `input_length`.

shamb0 replied:
Thanks for your comments. I picked up some useful coding practices.

b41sh commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +240,123 @@ pub fn register(registry: &mut FunctionRegistry) {
+
+                                let inner_builder_type = match input_map_sref.infer_data_type() {
+                                    DataType::Map(box typ) => typ,
+                                    DataType::EmptyMap => DataType::EmptyMap,
```
this can be removed, as the `ScalarRef::Map(col)` can't have `EmptyMap` type.

shamb0 replied:
I have a quick clarification regarding the test cases for the EmptyMap condition. I have captured the current behavior of the function as follows:

```rust
// Deleting keys from an empty map
run_ast(file, "map_delete({})", &[]);

run_ast(file, "map_delete({}, NULL, NULL)", &[]);

// Deleting keys from a map literal
run_ast(file, "map_delete({}, ['k3'], ['k2'])", &[]);
run_ast(
    file,
    "map_delete({'k1': 'v1', 'k2': 'v2', 'k3': 'v3', 'k4': 'v4'}, 'k3', 'k2')",
    &[],
);
```

Here are the unit test results:

```shell
error: 
  --> SQL:1:1
  |
1 | map_delete({})
  | ^^^^^^^^^^^^^^ no overload satisfies `map_delete(Map(Nothing))`


error: 
  --> SQL:1:1
  |
1 | map_delete({}, NULL, NULL)
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^ no overload satisfies `map_delete(Map(Nothing), NULL, NULL)`


error: 
  --> SQL:1:1
  |
1 | map_delete({}, ['k3'], ['k2'])
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no overload satisfies `map_delete(Map(Nothing), Array(String), Array(String))`
```

Is this behavior aligned with the requirements? I couldn't find details for handling the EmptyMap condition in the documentation: [Snowflake Documentation](https://docs.snowflake.com/en/sql-reference/functions/map_delete).

b41sh replied:
These sqls do not match the type required by the `map_delete` function. The first example has only one argument, and the second and third examples have the wrong type of key value, so return the error is correct.

b41sh commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +240,123 @@ pub fn register(registry: &mut FunctionRegistry) {
+
+                                let inner_builder_type = match input_map_sref.infer_data_type() {
+                                    DataType::Map(box typ) => typ,
+                                    DataType::EmptyMap => DataType::EmptyMap,
```
this can be removed, as the `ScalarRef::Map(col)` can't have `EmptyMap` type.

b41sh commented on src/query/functions/src/scalars/map.rs:
```diff
@@ -229,6 +240,114 @@ pub fn register(registry: &mut FunctionRegistry) {
+        }
+
+        let map_key_type = match args_type[0].remove_nullable() {
+            DataType::Map(box DataType::Tuple(type_tuple)) if type_tuple.len() == 2 => {
```
We need support `EmptyMap` here, the previous code is correct, you can modify it back. Delete `DataType::EmptyMap` on line 317 is because we have already processed the `EmptyMap` type on line 299, so `EmptyMap` type can't appear on line 317. 
I think there is basically ok with the rest of the codes. After this modification is completed, we can merge this PR.

shamb0 replied:
Thankyou

b41sh replied:
Hi @shamb0 , seems not correct. `select map_delete({}, 'a')` will return `SemanticError`, you can write like this:
```rust
let map_key_type = match args_type[0].remove_nullable() {
    DataType::Map(box DataType::Tuple(type_tuple)) if type_tuple.len() == 2 => {
        Some(type_tuple[0].clone())
    }
    DataType::EmptyMap => None,
    _ => return None,
};

if let Some(map_key_type) = map_key_type {
    for arg_type in args_type.iter().skip(1) {
        if arg_type != &map_key_type {
            return None;
        }
    }
} else {
    let key_type = &args_type[1];
    if !key_type.is_boolean()
        && !key_type.is_string()
        && !key_type.is_numeric()
        && !key_type.is_decimal()
        && !key_type.is_date_or_date_time()
    {
        return None;
    }
    for arg_type in args_type.iter().skip(2) {
        if arg_type != key_type {
            return None;
        }
    }
}
```
Please also remove `DataType::EmptyMap => DataType::EmptyMap` at line 325, because this case will never matched.

Add a unit test for `EmptyMap`, like `map_delete({}, 'a', 'b')`. After these changes are complete, we can merge this PR.

shamb0 replied:
Hi @b41sh,

Changes are done, and I've added the suggested test case. Could you please double-check the result? It looks correct on my end.

```txt
ast            : map_delete({}, 'a', 'b')
raw expr       : map_delete(map(array(), array()), 'a', 'b')
checked expr   : map_delete<Map(Nothing), String, String>(map<Array(Nothing), Array(Nothing)>(array<>(), array<>()), "a", "b")
optimized expr : {} :: Map(Nothing)
output type    : Map(Nothing)
output domain  : {}
output         : {}
```

b41sh reviewed (approved):
LGTM @shamb0 Thank you for your continued contribution over the last few days.
